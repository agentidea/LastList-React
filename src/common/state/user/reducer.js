import { getJwt } from './utils/jwt'
import {
  LOGIN_SUCCESSFULL,
  SIGN_OUT,
  SET_CURRENT_USER,
  PAY_API,
  RESET_PASSWORD_SUCCESSFULL,
} from './actions'

const initialState = {
  jwt: getJwt(),
  _id: null,
  email: null,
  firstName: null,
  lastName: null,
  dob: null,
  states: [],
  flow: null,
  payments: [],
  success: null,
}

export default (state = initialState, action) => {
  switch (action.type) {
    case LOGIN_SUCCESSFULL:
      return {
        ...state,
        jwt: action.jwt,
        ...action.data,
      }
    case SIGN_OUT:
      return {
        ...initialState,
        jwt: null,
      }
    case SET_CURRENT_USER:
      return {
        ...state,
        ...action.data,
      }
    case PAY_API.SUCCESS:
      return {
        ...state,
        states: action.data.user.states,
        payments: action.data.user.payments,
      }
    case RESET_PASSWORD_SUCCESSFULL:
      return {
        ...state,
        success: action.data,
      }
    default: {
      return state
    }
  }
}
