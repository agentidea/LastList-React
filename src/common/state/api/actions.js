import apiRequest from '../../utils/api'
import { jwtSelector } from '../user/selectors'

export function createAPIActions(prefix, suffix) {
  const PREFIX = prefix.toUpperCase()
  const SUFFIX = suffix.toUpperCase()
  return {
    REQUEST: `${PREFIX}/${SUFFIX}_REQUEST`,
    SUCCESS: `${PREFIX}/${SUFFIX}_SUCCESS`,
    FAILURE: `${PREFIX}/${SUFFIX}_FAILURE`,
  }
}

export function doRequest(actions, endpoint, body) {
  return async (dispatch, getState) => {
    const jwt = jwtSelector(getState())
    dispatch({
      type: actions.REQUEST,
      endpoint,
      body,
      jwt,
      meta: body ? body.body : null,
    })

    try {
      const data = await apiRequest(endpoint, jwt, body)

      dispatch({
        type: actions.SUCCESS,
        data,
        meta: body ? body.body : null,
      })

      return data
    } catch (error) {
      dispatch({
        type: actions.FAILURE,
        error,
        meta: body ? body.body : null,
      })

      throw error
    }
  }
}
