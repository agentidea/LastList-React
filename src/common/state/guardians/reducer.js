import { ADD_GUARDIAN_API, GET_GUARDIANS_API } from './actions'

const initialState = {
  loading: false,
  guardians: [],
}

export default (state = initialState, action) => {
  switch (action.type) {
    case GET_GUARDIANS_API.REQUEST:
      return {
        ...state,
        loading: true,
      }
    case GET_GUARDIANS_API.SUCCESS:
      return {
        ...state,
        loading: false,
        guardians: action.data,
      }
    case GET_GUARDIANS_API.FAILURE:
      return {
        ...state,
        loading: false,
      }
    case ADD_GUARDIAN_API.SUCCESS:
      return {
        ...state,
        guardians: action.data,
      }
    default: {
      return state
    }
  }
}
