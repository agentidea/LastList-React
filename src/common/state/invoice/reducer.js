import { GET_INVOICE_API } from './actions'

const initialState = {
  loading: false,
  error: null,
  data: {
    totalSongs: 0,
    listCost: 0,
    due: 0,
    numPayments: 0,
    totalLists: 0,
  },
}

export default (state = initialState, action) => {
  switch (action.type) {
    case GET_INVOICE_API.REQUEST:
      return {
        ...state,
        loading: true,
        error: null,
      }
    case GET_INVOICE_API.SUCCESS:
      return {
        ...state,
        loading: false,
        data: action.data.invoice,
      }
    case GET_INVOICE_API.FAIlURE:
      return {
        ...state,
        error: 'some error',
        loading: false,
      }
    default: {
      return state
    }
  }
}
