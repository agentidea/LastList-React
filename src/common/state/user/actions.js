import store from 'store/dist/store.modern'
import { createAPIActions, doRequest } from '../api/actions'
import { resetJwt, setJwt } from './utils/jwt'
import { jwtSelector, currentUserSelector } from './selectors'
import UserError from './error'

export const LOGIN_REG_SUCCESSFULL = 'LOGIN_REGISTRATION_SUCCESSFULL'
export const LOGIN_SUCCESSFULL = 'LOGIN_SUCCESSFULL'
export const RESET_PASSWORD_SUCCESSFULL = 'RESET_PASSWORD_SUCCESSFULL'
export const SET_CURRENT_USER = 'SET_CURRENT_USER'
export const SIGN_OUT = 'SIGN_OUT'

export const SIGN_OUT_API = createAPIActions('SIGN_OUT_API', 'DELETE')
export const GET_CURRENT_USER_API = createAPIActions('CURRENT_USER', 'FETCH')
export const LOGIN_API = createAPIActions('LOGIN_API', 'POST')
export const SIGNUP_API = createAPIActions('SIGNUP_API', 'POST')
export const CONFIRM_USER_API = createAPIActions('CONFIRM_USER_API', 'POST')
export const FORGOT_PASSWORD_API = createAPIActions('FORGOT_PASSWORD_API', 'POST')
export const RESET_PASSWORD_API = createAPIActions('RESET_PASSWORD_API', 'POST')
export const SAVE_PROFILE_API = createAPIActions('SAVE_PROFILE_API', 'PUT')
export const PAY_API = createAPIActions('PAY_API', 'POST')

export const signOut = () => async dispatch => {
  await resetJwt()
  store.set('jwtEmail', null)
  dispatch({
    type: SIGN_OUT,
  })
}

export const getCurrentUser = () => async (dispatch, getState) => {
  const jwt = jwtSelector(getState())
  if (jwt) {
    try {
      // const { data } = await dispatch(doRequest(GET_CURRENT_USER_API, `user/current`))
      // this is a fake call to simulate a /user/current while it's not available
      const user = await dispatch(
        doRequest(GET_CURRENT_USER_API, `user/validate`, {
          method: 'POST',
          body: {
            email: store.get('jwtEmail'),
            jwt: jwt,
          },
        })
      )

      dispatch({ type: SET_CURRENT_USER, data: user })
    } catch (e) {
      if (e.response && e.response.status === 401) {
        await dispatch(signOut())
      } else {
        throw e
      }
    }
  }
}

export const login = (email, password) => async dispatch => {
  try {
    const user = await dispatch(
      doRequest(LOGIN_API, `user/login`, {
        method: 'POST',
        body: { email, password },
      })
    )
    const jwt = user.jwt
    if (jwt) {
      setJwt(jwt)
      store.set('jwtEmail', email)
      dispatch({
        type: LOGIN_SUCCESSFULL,
        jwt,
        data: user,
      })
      return dispatch({ type: SET_CURRENT_USER, data: user })
    }
  } catch (e) {
    if (e.response && e.response.status === 409) {
      const message = e.response.data.message || 'User not found.'
      throw new UserError(message, 'email')
    }
    throw e
  }
}

export const signup = (email, password) => async dispatch => {
  try {
    const result = await dispatch(
      doRequest(SIGNUP_API, `user/signup`, {
        method: 'POST',
        body: { email, password },
      })
    )
    return result
  } catch (e) {
    if (e.response && e.response.data) {
      const message = e.response.data.message
      throw new UserError(message || 'An error occurred.', 'email')
    }
    throw e
  }
}

// export const confirmAccount = confirmationCode => async dispatch => {
//   try {
//     const user = await dispatch(
//       doRequest(CONFIRM_USER_API, `user/confirmEmail/${confirmationCode}`)
//     )
//     return user
//   } catch (e) {
//     if (e.response && e.response.data) {
//       const message = e.response.data.message
//       throw new UserError(message)
//     }
//     throw e
//   }
// }
export const confirmAccount = confirmationCode => async dispatch => {
  try {
    const user = await dispatch(
      doRequest(CONFIRM_USER_API, `user/confirmEmail/${confirmationCode}`)
    )
    const jwt = user._id // todo replace this with real JWT
    const email = user.email //temp solution for now
    if (jwt) {
      setJwt(jwt)
      store.set('jwtEmail', email) // temporary until we have real JWT
      dispatch({ type: LOGIN_SUCCESSFULL, jwt, data: user })
      return dispatch({ type: SET_CURRENT_USER, data: user })
    }
  } catch (e) {
    if (e.response && e.response.status === 409) {
      const message = e.response.data.message || 'User not found.'
      throw new UserError(message, 'email')
    }
    throw e
  }
}

export const saveUserProfile = (firstName, lastName, dob) => async (dispatch, getState) => {
  const user = currentUserSelector(getState())
  if (user) {
    try {
      const result = await dispatch(
        doRequest(SAVE_PROFILE_API, `user/profile/${user._id}`, {
          method: 'PUT',
          body: {
            firstName,
            lastName,
            dob: dob.toISOString(),
          },
        })
      )
      return result
    } catch (e) {
      if (e.response && e.response.status === 409) {
        const message = e.response.data.message || 'Unknown error'
        const field = e.response.data.field
        throw new UserError(message, field)
      }
      throw e
    }
  } else {
    console.error('User should be logged in to save his profile')
  }
}

export const payForList = (amount, paymentMethod, numberListsPayingFor) => async (
  dispatch,
  getState
) => {
  const user = currentUserSelector(getState())
  if (user) {
    await dispatch(
      doRequest(PAY_API, `user/payment/${user._id}`, {
        method: 'POST',
        body: {
          amount,
          paymentMethod,
          numberListsPayingFor,
        },
      })
    )
    return true
  } else {
    console.error('User should be logged in to pay for lists')
  }
}

export const forgotPassword = resetToken => async dispatch => {
  try {
    const user = await dispatch(
      doRequest(FORGOT_PASSWORD_API, `user/forgot-password`, {
        method: 'POST',
        resetToken: resetToken,
      })
    )
    const code = user.code
    if (code === 200) {
      return true
    }
  } catch (e) {
    if (e.response) {
      const message = e.response.data.message || 'Email provided can not be found!'
      throw new UserError(message, 'email')
    }
    throw e
  }
}

export const resetPassword = password => async dispatch => {
  try {
    const user = await dispatch(
      doRequest(RESET_PASSWORD_API, `user/reset-password`, {
        method: 'POST',
        resetToken: password,
      })
    )
    const code = user.code
    if (code === 200) {
      let success = 'Your Password was successfully reset, now you can login with your new one'
      dispatch({ type: RESET_PASSWORD_SUCCESSFULL, data: success })
      return true
    }
  } catch (e) {
    if (e.response) {
      const message = e.response.data.message || 'Something went wrong!'
      throw new UserError(message, 'email')
    }
    throw e
  }
}
