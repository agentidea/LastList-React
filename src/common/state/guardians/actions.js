import { createAPIActions, doRequest } from '../api/actions'
import { currentUserSelector } from '../user/selectors'

export const GET_GUARDIANS_API = createAPIActions('GET_GUARDIANS_API', 'FETCH')
export const ADD_GUARDIAN_API = createAPIActions('ADD_GUARDIAN_API', 'POST')
export const REMOVE_GUARDIAN_API = createAPIActions('REMOVE_GUARDIAN_API', 'DELETE')

export const fetchUserGuardians = () => async (dispatch, getState) => {
  const user = currentUserSelector(getState())
  if (user) {
    await dispatch(doRequest(GET_GUARDIANS_API, `user/getGuaridans/${user._id}`))
  } else {
    console.error('User should be logged in to fetch his guardians')
  }
}

export const addGuardian = (firstName, lastName, email) => async (dispatch, getState) => {
  const user = currentUserSelector(getState())
  if (user) {
    await dispatch(
      doRequest(ADD_GUARDIAN_API, `user/guardian/${user._id}`, {
        method: 'POST',
        body: {
          firstName,
          lastName,
          email,
        },
      })
    )
  } else {
    console.error('User should be logged in to add a guardian')
  }
}

export const removeGuardian = uuid => async (dispatch, getState) => {
  const user = currentUserSelector(getState())
  if (user) {
    await dispatch(
      doRequest(REMOVE_GUARDIAN_API, `user/guardian/${user._id}`, {
        method: 'DELETE',
        body: { uuid },
      })
    )
  } else {
    console.error('User should be logged in to remove a guardian')
  }
}
