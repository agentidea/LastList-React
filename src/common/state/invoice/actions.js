import { createAPIActions, doRequest } from '../api/actions'
import { currentUserSelector } from '../user/selectors'

export const GET_INVOICE_API = createAPIActions('GET_INVOICE_API', 'FETCH')

export const getInvoice = () => async (dispatch, getState) => {
  const user = currentUserSelector(getState())
  if (user) {
    dispatch(doRequest(GET_INVOICE_API, `user/invoice/${user._id}`))
  } else {
    console.error('User should be logged in to get the invoice')
  }
}
