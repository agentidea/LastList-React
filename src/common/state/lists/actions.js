import { createAPIActions, doRequest } from '../api/actions'
import { listsForServerSelector } from './selectors'

import { currentUserSelector } from '../user/selectors'
import getListsData from '../mocks/getLists'

export const GET_LISTS_API = createAPIActions('GET_LISTS_API', 'FETCH')
export const SAVE_LIST_API = createAPIActions('SAVE_LIST_API', 'PUT')

//export const INVOICE_LIST_API = createAPIActions('INVOICE_LIST_API', 'GET')

export const ADD_NEW_LIST = 'ADD_NEW_LIST'
export const UPDATE_LIST_FIELD = 'UPDATE_LIST_FIELD'

const fakeRequest = (dispatch, action, user, data) => {
  return new Promise(resolve => {
    dispatch({ type: action.REQUEST, userId: user._id })
    setTimeout(() => {
      dispatch({ type: action.SUCCESS, data })
      resolve()
    }, 500)
  })
}

// export const fetchUserListInvoice = () => async (dispatch, getState) => {
//   const user = currentUserSelector(getState())
//   if (user) {
//     await dispatch(doRequest(INVOICE_LIST_API, `user/invoice/${user._id}`))
//   } else {
//     console.error('User should be logged in to fetch his invoice')
//   }
// }

export const fetchUserLists = () => async (dispatch, getState) => {
  const user = currentUserSelector(getState())
  if (user) {
    // await fakeRequest(dispatch, GET_LISTS_API, user, getListsData)
    await dispatch(doRequest(GET_LISTS_API, `user/getLists/${user._id}`))
  } else {
    console.error('User should be logged in to fetch his lists')
  }
}

export const setListItemArtist = (listIndex, itemIndex, value) => ({
  type: UPDATE_LIST_FIELD,
  field: 'artistName',
  listIndex,
  itemIndex,
  value,
})

export const setListItemSong = (listIndex, itemIndex, value) => ({
  type: UPDATE_LIST_FIELD,
  field: 'songName',
  listIndex,
  itemIndex,
  value,
})

export const saveUserList = () => async (dispatch, getState) => {
  const user = currentUserSelector(getState())
  if (user) {
    await dispatch(
      doRequest(SAVE_LIST_API, `user/addList/${user._id}`, {
        method: 'PUT',
        body: {
          sets: listsForServerSelector(getState()),
        },
      })
    )
  } else {
    console.error('User should be logged in to save his lists')
  }
}

export const addNewList = () => ({
  type: ADD_NEW_LIST,
})
