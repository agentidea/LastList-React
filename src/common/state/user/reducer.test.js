import reducer from './reducer.js'
import { LOGIN_SUCCESSFULL, SIGN_OUT, SET_CURRENT_USER } from './actions'

describe('User reducer ', () => {
  test('default state should contain correct data', () => {
    const state = reducer(undefined, { type: '' })
    expect(state).toMatchSnapshot()
  })
  test('LOGIN_SUCCESSFULL should set JWT', () => {
    const state = reducer({ jwt: null }, { type: LOGIN_SUCCESSFULL, jwt: 'abc' })
    expect(state).toMatchSnapshot()
  })
  test('SIGN_OUT should unset the JWT and reset the user data', () => {
    const state = reducer({ jwt: 'def' }, { type: SIGN_OUT })
    expect(state).toMatchSnapshot()
  })
  test('SET_CURRENT_USER should unset the JWT and reset the user data', () => {
    const user = {
      _id: 'XYZ',
      email: 'jon.snow@gmail.com',
      firstName: 'Jon',
      lastName: 'Snow',
    }
    const state = reducer({ jwt: 'abc' }, { type: SET_CURRENT_USER, data: user })
    expect(state).toMatchSnapshot()
  })
})
