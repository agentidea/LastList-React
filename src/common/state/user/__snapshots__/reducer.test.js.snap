// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User reducer  LOGIN_SUCCESSFULL should set JWT 1`] = `
Object {
  "jwt": "abc",
}
`;

exports[`User reducer  SET_CURRENT_USER should unset the JWT and reset the user data 1`] = `
Object {
  "_id": "XYZ",
  "email": "jon.snow@gmail.com",
  "firstName": "Jon",
  "jwt": "abc",
  "lastName": "Snow",
  "loading": false,
}
`;

exports[`User reducer  SIGN_OUT should unset the JWT and reset the user data 1`] = `
Object {
  "_id": null,
  "dob": null,
  "email": null,
  "firstName": null,
  "jwt": null,
  "lastName": null,
  "loading": true,
  "payments": Array [],
  "states": null,
}
`;

exports[`User reducer  default state should contain correct data 1`] = `
Object {
  "_id": null,
  "dob": null,
  "email": null,
  "firstName": null,
  "jwt": undefined,
  "lastName": null,
  "loading": true,
  "payments": Array [],
  "states": null,
}
`;
